File: gospider/examples/ecommerce/detector.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"strings"
)

// ═══════════════════════════════════════════════════════════════════════
// Universal E-commerce Site Detector (Go version)
// Auto-detects any e-commerce site from URL patterns, HTML structure,
// JSON-LD metadata, and known platform signatures.
// ═══════════════════════════════════════════════════════════════════════

// EcommerceDetectionResult holds the outcome of e-commerce site detection.
type EcommerceDetectionResult struct {
	IsEcommerce      bool     `json:"is_ecommerce"`
	Confidence       float64  `json:"confidence"`
	SiteFamily       string   `json:"site_family"`
	Platform         string   `json:"platform"`
	DetectedFeatures []string `json:"detected_features"`
	Currency         string   `json:"currency"`
	HasJSONLD        bool     `json:"has_jsonld"`
	HasNextData      bool     `json:"has_next_data"`
	PriceAPIDetected bool     `json:"price_api_detected"`
	CartURL          string   `json:"cart_url"`
	CategoryURLs     []string `json:"category_urls"`
}

// URL_SIGNATURES maps platform names to their URL regex patterns and confidence.
var URL_SIGNATURES = map[string]map[string]interface{}{
	// China
	"jd":            {"patterns": []string{`jd\.com`, `jd\.hk`}, "confidence": 0.95},
	"taobao":        {"patterns": []string{`taobao\.com`}, "confidence": 0.95},
	"tmall":         {"patterns": []string{`tmall\.com`}, "confidence": 0.95},
	"pinduoduo":     {"patterns": []string{`pinduoduo\.com`, `yangkeduo\.com`, `pdd\.com`}, "confidence": 0.95},
	"1688":          {"patterns": []string{`1688\.com`}, "confidence": 0.95},
	"suning":        {"patterns": []string{`suning\.com`}, "confidence": 0.90},
	"vip":           {"patterns": []string{`vip\.com`, `vipshop\.com`}, "confidence": 0.95},
	"dangdang":      {"patterns": []string{`dangdang\.com`}, "confidence": 0.95},
	"xiaohongshu":   {"patterns": []string{`xiaohongshu\.com`, `xhscdn\.com`}, "confidence": 0.95},
	"douyin-shop":   {"patterns": []string{`douyin\.com`, `jinritemai\.com`}, "confidence": 0.90},
	"kuaishou-shop": {"patterns": []string{`kuaishou\.com`, `kwai\.com`}, "confidence": 0.90},
	"kaola":         {"patterns": []string{`kaola\.com`}, "confidence": 0.95},
	"gome":          {"patterns": []string{`gome\.com\.cn`}, "confidence": 0.95},
	// International
	"amazon":       {"patterns": []string{`amazon\.(com|co\.uk|de|fr|it|es|co\.jp|com\.au|ca|in)`}, "confidence": 0.95},
	"ebay":         {"patterns": []string{`ebay\.(com|co\.uk|de|fr)`}, "confidence": 0.95},
	"aliexpress":   {"patterns": []string{`aliexpress\.(com|us|es|ru)`}, "confidence": 0.95},
	"lazada":       {"patterns": []string{`lazada\.(com|co|com\.my|com\.ph|co\.id|co\.th|vn)`}, "confidence": 0.95},
	"shopee":       {"patterns": []string{`shopee\.(com|co\.th|co\.id|com\.my|com\.ph|vn|tw|br)`}, "confidence": 0.95},
	"rakuten":      {"patterns": []string{`rakuten\.(co\.jp|com)`}, "confidence": 0.95},
	"walmart":      {"patterns": []string{`walmart\.(com|ca)`}, "confidence": 0.95},
	"bestbuy":      {"patterns": []string{`bestbuy\.(com|ca)`}, "confidence": 0.95},
	"target":       {"patterns": []string{`target\.com`}, "confidence": 0.95},
	"costco":       {"patterns": []string{`costco\.(com|ca)`}, "confidence": 0.95},
	"newegg":       {"patterns": []string{`newegg\.(com|ca)`}, "confidence": 0.95},
	"temu":         {"patterns": []string{`temu\.(com|co\.uk|co\.jp|de|fr)`}, "confidence": 0.95},
	"shein":        {"patterns": []string{`shein\.(com|co\.uk|co\.jp|de|fr)`}, "confidence": 0.95},
	"wish":         {"patterns": []string{`wish\.com`}, "confidence": 0.95},
	"mercadolibre": {"patterns": []string{`mercadolibre\.(com\.ar|com\.mx|com\.br)`}, "confidence": 0.95},
	"ozon":         {"patterns": []string{`ozon\.ru`}, "confidence": 0.95},
	"wildberries":  {"patterns": []string{`wildberries\.ru`}, "confidence": 0.95},
	"allegro":      {"patterns": []string{`allegro\.(pl|cz|sk)`}, "confidence": 0.95},
	"cdiscount":    {"patterns": []string{`cdiscount\.com`}, "confidence": 0.95},
}

// PLATFORM_SIGNATURES detects CMS/SaaS e-commerce platforms.
var PLATFORM_SIGNATURES = map[string]map[string]interface{}{
	"shopify":     {"html": []string{"shopify.com", "cdn.shopify", "Shopify.theme"}, "confidence": 0.90},
	"magento":     {"html": []string{"Magento_", "mage-cache", "mage/cookies"}, "confidence": 0.90},
	"woocommerce": {"html": []string{"woocommerce", "wp-content/plugins/woocommerce"}, "confidence": 0.85},
	"bigcommerce": {"html": []string{"bigcommerce", "bc.js"}, "confidence": 0.90},
	"prestashop":  {"html": []string{"prestashop", "ps-shoppingcart"}, "confidence": 0.90},
	"wix":         {"html": []string{"wix.com", "wixstores"}, "confidence": 0.80},
	"squarespace": {"html": []string{"squarespace.com", "sqs-shop"}, "confidence": 0.80},
	"salesforce":  {"html": []string{"demandware", "sfcc", "commercecloud"}, "confidence": 0.85},
}

// ECOMMERCE_HTML_SIGNALS identifies e-commerce structural indicators.
var ECOMMERCE_HTML_SIGNALS = []string{
	`application/ld\+json[^"]*Product`,
	`"@type"\s*:\s*"Product"`,
	`"@type"\s*:\s*"Offer"`,
	`"@type"\s*:\s*"AggregateOffer"`,
	`class="[^"]*price[^"]*"`,
	`class="[^"]*add-to-cart[^"]*"`,
	`class="[^"]*buy-now[^"]*"`,
	`class="[^"]*add_to_cart[^"]*"`,
	`shopping[\-]?cart`,
	`data-product-id`,
	`data-sku`,
	`data-variant-id`,
	`itemtype="https?://schema\.org/Product"`,
	`[\$\€\£\¥\₹][\d,]+\.?\d*`,
	`\uffe5[\d,]+`,
}

// DetectEcommerceSite detects if a URL/HTML represents an e-commerce site.
func DetectEcommerceSite(rawURL string, html string) EcommerceDetectionResult {
	result := EcommerceDetectionResult{
		SiteFamily: "generic",
	}

	if rawURL == "" {
		return result
	}

	// Layer 1: URL pattern matching (fast, high confidence)
	urlMatch := detectByURL(rawURL)
	if urlMatch.IsEcommerce {
		result = urlMatch
	}

	// Layer 2: Platform/CMS signature detection
	if !result.IsEcommerce && html != "" {
		platformMatch := detectByPlatform(html)
		if platformMatch.IsEcommerce {
			result = platformMatch
		}
	}

	// Layer 3: HTML structure analysis
	if !result.IsEcommerce && html != "" {
		htmlMatch := detectByHTMLSignals(html)
		if htmlMatch.IsEcommerce {
			result = htmlMatch
		}
		// Merge detected features
		result.DetectedFeatures = append(result.DetectedFeatures, htmlMatch.DetectedFeatures...)
	}

	// Layer 4: Embedded structured data parsing
	if html != "" {
		if strings.Contains(html, "application/ld+json") {
			result.HasJSONLD = true
			result.DetectedFeatures = append(result.DetectedFeatures, "json_ld")
		}
		if strings.Contains(html, "__NEXT_DATA__") {
			result.HasNextData = true
			result.DetectedFeatures = append(result.DetectedFeatures, "next_data")
		}
		if !result.IsEcommerce && (result.HasJSONLD || result.HasNextData) {
			products := extractDetectorJSONLDProducts(html)
			if len(products) > 0 {
				result.IsEcommerce = true
				result.Confidence = 0.70
				result.DetectedFeatures = append(result.DetectedFeatures, "embedded_product_data")
			}
		}
	}

	return result
}

func detectByURL(rawURL string) EcommerceDetectionResult {
	for family, sig := range URL_SIGNATURES {
		patterns, ok := sig["patterns"].([]string)
		if !ok {
			continue
		}
		for _, pat := range patterns {
			re := regexp.MustCompile(pat)
			if re.MatchString(rawURL) {
				confidence := 0.95
				if c, ok := sig["confidence"].(float64); ok {
					confidence = c
				}
				return EcommerceDetectionResult{
					IsEcommerce:      true,
					Confidence:       confidence,
					SiteFamily:       family,
					DetectedFeatures: []string{"url_pattern"},
				}
			}
		}
	}
	return EcommerceDetectionResult{}
}

func detectByPlatform(html string) EcommerceDetectionResult {
	for platform, sig := range PLATFORM_SIGNATURES {
		htmlSigs, ok := sig["html"].([]string)
		if !ok {
			continue
		}
		for _, needle := range htmlSigs {
			if strings.Contains(html, needle) {
				confidence := 0.85
				if c, ok := sig["confidence"].(float64); ok {
					confidence = c
				}
				return EcommerceDetectionResult{
					IsEcommerce:      true,
					Confidence:       confidence,
					SiteFamily:       platform,
					Platform:         platform,
					DetectedFeatures: []string{"platform_signature"},
				}
			}
		}
	}
	return EcommerceDetectionResult{}
}

// detectByHTMLSignals analyzes HTML structure for e-commerce signals.
func detectByHTMLSignals(html string) EcommerceDetectionResult {
	signals := 0
	hasJsonld := false
	hasNextData := false

	for _, signal := range ECOMMERCE_HTML_SIGNALS {
		re := regexp.MustCompile(signal)
		if re.MatchString(html) {
			signals++
		}
	}

	if strings.Contains(html, "application/ld+json") {
		hasJsonld = true
		signals += 2
	}
	if strings.Contains(html, "__NEXT_DATA__") || strings.Contains(html, "__INITIAL_STATE__") {
		hasNextData = true
		signals++
	}

	confidence := math.Min(0.8, float64(signals)*0.15)
	if signals >= 2 {
		return EcommerceDetectionResult{
			IsEcommerce:      true,
			Confidence:       confidence,
			HasJSONLD:        hasJsonld,
			HasNextData:      hasNextData,
			DetectedFeatures: []string{"html_structure"},
		}
	}
	return EcommerceDetectionResult{}
}

// EXPECTED_JSONLD_FIELDS lists product-related schema.org fields.
var EXPECTED_JSONLD_FIELDS = []string{
	"@type", "name", "image", "offers",
	"price", "priceCurrency", "availability",
	"sku", "brand", "review", "aggregateRating",
}

// detectByStructuredData looks for structured product data in script tags.
func detectByStructuredData(html string) EcommerceDetectionResult {
	for _, field := range EXPECTED_JSONLD_FIELDS {
		if strings.Contains(html, fmt.Sprintf("\"%s\"", field)) {
			return EcommerceDetectionResult{
				IsEcommerce:      true,
				Confidence:       0.6,
				HasJSONLD:        true,
				DetectedFeatures: []string{"structured_data"},
			}
		}
	}
	return EcommerceDetectionResult{}
}

// detectByPriceHeuristic detects price-related patterns in the page.
func detectByPriceHeuristic(html string) EcommerceDetectionResult {
	priceIndicators := []string{
		"price", "amount", "currency",
		"discount", "originalPrice", "salePrice",
	}
	hits := 0
	for _, ind := range priceIndicators {
		if strings.Contains(html, ind) {
			hits++
		}
	}
	if hits >= 3 {
		return EcommerceDetectionResult{
			IsEcommerce:      true,
			Confidence:       0.5,
			PriceAPIDetected: true,
			DetectedFeatures: []string{"price_api"},
		}
	}
	return EcommerceDetectionResult{}
}

// extractJSONLDProducts extracts Product-type JSON-LD blocks from HTML.
func extractDetectorJSONLDProducts(html string) []map[string]interface{} {
	re := regexp.MustCompile(`(?s)<script[^>]*type=\"application/ld\+json\"[^>]*>(.*?)</script>`)
	matches := re.FindAllStringSubmatch(html, -1)
	var products []map[string]interface{}
	for _, m := range matches {
		var data interface{}
		if err := json.Unmarshal([]byte(m[1]), &data); err != nil {
			continue
		}
		// Flatten @graph arrays
		var items []interface{}
		switch v := data.(type) {
		case map[string]interface{}:
			if graph, ok := v["@graph"].([]interface{}); ok {
				items = graph
			} else {
				items = []interface{}{v}
			}
		case []interface{}:
			items = v
		}
		for _, item := range items {
			obj, ok := item.(map[string]interface{})
			if !ok {
				continue
			}
			if t, ok := obj["@type"].(string); ok && (t == "Product" || t == "Offer" || t == "AggregateOffer") {
				products = append(products, obj)
			}
		}
	}
	return products
}

File: rustspider/examples/ecommerce/detector.rs
//! Universal e-commerce site detector for the Rust example crawler.

#![allow(dead_code)]

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EcommerceDetectionResult {
    pub is_ecommerce: bool,
    pub confidence: f64,
    pub site_family: String,
    pub platform: String,
    pub detected_features: Vec<String>,
    pub currency: String,
    pub has_jsonld: bool,
    pub has_next_data: bool,
    pub price_api_detected: bool,
    pub cart_url: String,
    pub category_urls: Vec<String>,
}

impl Default for EcommerceDetectionResult {
    fn default() -> Self {
        Self {
            is_ecommerce: false,
            confidence: 0.0,
            site_family: "generic".to_string(),
            platform: String::new(),
            detected_features: Vec::new(),
            currency: String::new(),
            has_jsonld: false,
            has_next_data: false,
            price_api_detected: false,
            cart_url: String::new(),
            category_urls: Vec::new(),
        }
    }
}

pub fn detect_ecommerce_site(raw_url: &str, html: &str) -> EcommerceDetectionResult {
    let mut result = EcommerceDetectionResult::default();
    if raw_url.trim().is_empty() {
        return result;
    }

    merge_result(&mut result, detect_by_url(raw_url));
    merge_result(&mut result, detect_by_platform(html));
    merge_result(&mut result, detect_by_html_signals(html));
    merge_result(&mut result, detect_by_structured_data(html));
    merge_result(&mut result, detect_by_price_api(html));

    if html.contains("application/ld+json") {
        result.has_jsonld = true;
        push_feature(&mut result, "json_ld");
    }
    if html.contains("__NEXT_DATA__")
        || html.contains("__INITIAL_STATE__")
        || html.contains("__NUXT__")
    {
        result.has_next_data = true;
        push_feature(&mut result, "hydration_data");
    }
    if result.site_family != "generic" {
        if let Some((cart, categories)) = site_family_urls().get(result.site_family.as_str()) {
            result.cart_url = (*cart).to_string();
            result.category_urls = categories.iter().map(|url| (*url).to_string()).collect();
        }
    }
    result
}

fn merge_result(
    result: &mut EcommerceDetectionResult,
    candidate: Option<EcommerceDetectionResult>,
) {
    let Some(candidate) = candidate else {
        return;
    };
    result.is_ecommerce = result.is_ecommerce || candidate.is_ecommerce;
    if candidate.confidence > result.confidence {
        result.confidence = candidate.confidence;
    }
    if result.site_family == "generic" && candidate.site_family != "generic" {
        result.site_family = candidate.site_family;
    }
    if result.platform.is_empty() && !candidate.platform.is_empty() {
        result.platform = candidate.platform;
    }
    if result.currency.is_empty() && !candidate.currency.is_empty() {
        result.currency = candidate.currency;
    }
    result.has_jsonld |= candidate.has_jsonld;
    result.has_next_data |= candidate.has_next_data;
    result.price_api_detected |= candidate.price_api_detected;
    for feature in candidate.detected_features {
        push_feature(result, &feature);
    }
}

fn push_feature(result: &mut EcommerceDetectionResult, feature: &str) {
    if !result
        .detected_features
        .iter()
        .any(|value| value == feature)
    {
        result.detected_features.push(feature.to_string());
    }
}

fn detect_by_url(raw_url: &str) -> Option<EcommerceDetectionResult> {
    for (family, patterns, confidence, currency) in url_signatures() {
        if patterns.iter().any(|pattern| regex_match(pattern, raw_url)) {
            return Some(EcommerceDetectionResult {
                is_ecommerce: true,
                confidence,
                site_family: family.to_string(),
                platform: family.to_string(),
                currency: currency.to_string(),
                detected_features: vec!["url_pattern".to_string()],
                ..Default::default()
            });
        }
    }
    None
}

fn detect_by_platform(html: &str) -> Option<EcommerceDetectionResult> {
    if html.is_empty() {
        return None;
    }
    let lowered = html.to_ascii_lowercase();
    for (platform, needles, confidence) in platform_signatures() {
        if needles.iter().any(|needle| lowered.contains(needle)) {
            return Some(EcommerceDetectionResult {
                is_ecommerce: true,
                confidence,
                platform: platform.to_string(),
                detected_features: vec!["platform_signature".to_string()],
                ..Default::default()
            });
        }
    }
    None
}

fn detect_by_html_signals(html: &str) -> Option<EcommerceDetectionResult> {
    if html.is_empty() {
        return None;
    }
    let mut signals = 0;
    for pattern in html_signals() {
        if regex_match(pattern, html) {
            signals += 1;
        }
    }
    let has_jsonld = html.contains("application/ld+json");
    let has_next_data = html.contains("__NEXT_DATA__") || html.contains("__INITIAL_STATE__");
    if has_jsonld {
        signals += 2;
    }
    if has_next_data {
        signals += 1;
    }
    if signals < 2 {
        return None;
    }
    Some(EcommerceDetectionResult {
        is_ecommerce: true,
        confidence: (signals as f64 * 0.15).min(0.85),
        has_jsonld,
        has_next_data,
        detected_features: vec!["html_structure".to_string()],
        ..Default::default()
    })
}

fn detect_by_structured_data(html: &str) -> Option<EcommerceDetectionResult> {
    for payload in json_payloads(html) {
        if let Ok(value) = serde_json::from_str::<Value>(&payload) {
            if contains_product_shape(&value) {
                return Some(EcommerceDetectionResult {
                    is_ecommerce: true,
                    confidence: 0.75,
                    has_jsonld: true,
                    detected_features: vec!["structured_product_data".to_string()],
                    ..Default::default()
                });
            }
        }
    }
    None
}

fn detect_by_price_api(html: &str) -> Option<EcommerceDetectionResult> {
    let lowered = html.to_ascii_lowercase();
    let hits = [
        "price",
        "amount",
        "currency",
        "discount",
        "saleprice",
        "stock",
    ]
    .iter()
    .filter(|needle| lowered.contains(**needle))
    .count();
    if hits >= 3 {
        Some(EcommerceDetectionResult {
            is_ecommerce: true,
            confidence: 0.50,
            price_api_detected: true,
            detected_features: vec!["price_api".to_string()],
            ..Default::default()
        })
    } else {
        None
    }
}

fn contains_product_shape(value: &Value) -> bool {
    match value {
        Value::Object(map) => {
            let node_type = map
                .get("@type")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_ascii_lowercase();
            if ["product", "offer", "aggregateoffer", "itemlist"].contains(&node_type.as_str()) {
                return true;
            }
            let product_fields = ["sku", "price", "offers", "priceCurrency", "aggregateRating"];
            if product_fields
                .iter()
                .filter(|key| map.contains_key(**key))
                .count()
                >= 2
            {
                return true;
            }
            map.values().any(contains_product_shape)
        }
        Value::Array(values) => values.iter().any(contains_product_shape),
        _ => false,
    }
}

fn json_payloads(html: &str) -> Vec<String> {
    let mut values = Vec::new();
    for pattern in [
        r#"(?is)<script[^>]+type=["']application/ld\+json["'][^>]*>(.*?)</script>"#,
        r#"(?is)<script[^>]+type=["']application/json["'][^>]*>(.*?)</script>"#,
        r#"(?is)__NEXT_DATA__\s*=\s*(\{.*?\})\s*;"#,
        r#"(?is)__INITIAL_STATE__\s*=\s*(\{.*?\})\s*;"#,
    ] {
        if let Ok(regex) = Regex::new(pattern) {
            for capture in regex.captures_iter(html) {
                if let Some(payload) = capture.get(1) {
                    values.push(payload.as_str().trim().to_string());
                }
            }
        }
    }
    values
}

fn regex_match(pattern: &str, text: &str) -> bool {
    Regex::new(&format!("(?i){pattern}"))
        .map(|regex| regex.is_match(text))
        .unwrap_or(false)
}

fn url_signatures() -> Vec<(&'static str, Vec<&'static str>, f64, &'static str)> {
    vec![
        ("jd", vec![r"jd\.com", r"jd\.hk"], 0.95, "CNY"),
        ("taobao", vec![r"taobao\.com"], 0.95, "CNY"),
        ("tmall", vec![r"tmall\.com"], 0.95, "CNY"),
        (
            "pinduoduo",
            vec![r"pinduoduo\.com", r"yangkeduo\.com", r"pdd\.com"],
            0.95,
            "CNY",
        ),
        ("1688", vec![r"1688\.com"], 0.95, "CNY"),
        ("suning", vec![r"suning\.com"], 0.90, "CNY"),
        ("vip", vec![r"vip\.com", r"vipshop\.com"], 0.95, "CNY"),
        (
            "xiaohongshu",
            vec![r"xiaohongshu\.com", r"xhscdn\.com"],
            0.95,
            "CNY",
        ),
        (
            "douyin-shop",
            vec![r"douyin\.com", r"jinritemai\.com"],
            0.90,
            "CNY",
        ),
        (
            "amazon",
            vec![r"amazon\.(com|co\.uk|de|fr|it|es|co\.jp|com\.au|ca|in)"],
            0.95,
            "USD",
        ),
        (
            "ebay",
            vec![r"ebay\.(com|co\.uk|de|fr|it|es|com\.au|ca)"],
            0.95,
            "USD",
        ),
        (
            "aliexpress",
            vec![r"aliexpress\.(com|us|es|ru|pt|fr|de|it)"],
            0.95,
            "USD",
        ),
        (
            "shopee",
            vec![r"shopee\.(com|co\.th|co\.id|com\.my|com\.ph|vn|tw|br)"],
            0.95,
            "USD",
        ),
        (
            "lazada",
            vec![r"lazada\.(com|com\.my|com\.ph|co\.id|co\.th|vn)"],
            0.95,
            "USD",
        ),
        ("walmart", vec![r"walmart\.(com|ca)"], 0.95, "USD"),
        ("target", vec![r"target\.com"], 0.95, "USD"),
        (
            "temu",
            vec![r"temu\.(com|co\.uk|co\.jp|de|fr|es|it|nl)"],
            0.95,
            "USD",
        ),
        (
            "shein",
            vec![r"shein\.(com|co\.uk|co\.jp|de|fr|es|it|nl)"],
            0.95,
            "USD",
        ),
        (
            "mercadolibre",
            vec![r"mercadolibre\.(com\.ar|com\.mx|com\.br|com\.co|cl)"],
            0.95,
            "USD",
        ),
        ("ozon", vec![r"ozon\.ru"], 0.95, "RUB"),
        ("wildberries", vec![r"wildberries\.ru"], 0.95, "RUB"),
        ("allegro", vec![r"allegro\.(pl|cz|sk)"], 0.95, "PLN"),
    ]
}

fn platform_signatures() -> Vec<(&'static str, Vec<&'static str>, f64)> {
    vec![
        (
            "shopify",
            vec!["shopify.com", "cdn.shopify", "shopify.theme"],
            0.90,
        ),
        (
            "magento",
            vec!["magento_", "mage-cache", "mage/cookies"],
            0.90,
        ),
        (
            "woocommerce",
            vec!["woocommerce", "wp-content/plugins/woocommerce"],
            0.85,
        ),
        ("bigcommerce", vec!["bigcommerce", "bc.js"], 0.90),
        ("prestashop", vec!["prestashop", "ps-shoppingcart"], 0.90),
        (
            "salesforce",
            vec!["demandware", "sfcc", "commercecloud"],
            0.85,
        ),
        ("wix", vec!["wix.com", "wixstores"], 0.80),
    ]
}

fn html_signals() -> Vec<&'static str> {
    vec![
        r#""@type"\s*:\s*"Product""#,
        r#""@type"\s*:\s*"Offer""#,
        r#"class=["'][^"']*price[^"']*["']"#,
        r#"class=["'][^"']*product[^"']*["']"#,
        r#"class=["'][^"']*add-to-cart[^"']*["']"#,
        r#"shopping[\-]?cart"#,
        r#"data-product-id"#,
        r#"data-sku"#,
        r#"data-variant-id"#,
        r#"itemtype=["']https?://schema\.org/Product["']"#,
        r#"[\$€£¥₹₩₽][\d,]+\.?\d*"#,
    ]
}

fn site_family_urls() -> HashMap<&'static str, (&'static str, Vec<&'static str>)> {
    HashMap::from([
        (
            "jd",
            ("https://cart.jd.com/", vec!["https://channel.jd.com/"]),
        ),
        (
            "taobao",
            (
                "https://cart.taobao.com/",
                vec!["https://www.taobao.com/tbhome/"],
            ),
        ),
        (
            "tmall",
            ("https://cart.tmall.com/", vec!["https://www.tmall.com/"]),
        ),
        (
            "amazon",
            (
                "https://www.amazon.com/gp/cart/",
                vec!["https://www.amazon.com/best-sellers/"],
            ),
        ),
    ])
}
